README.md

```diff
@@ -1,3 +1,65 @@
+# Overview
+
+Find below the original documentation for DLIO.
+
+This fork of DLIO was extended/modified with the objective of simulating as closely as possible the MLCommons Image Segmentation (UNET3D), BERT and DLRM workloads, and plug into our plotting software. This is still a work in progress.
+
+For this, we have added the following features:
+- Dockerfile with dependencies
+- support for evaluation phases on a held-out dataset
+- support for PyTorch DataLoader reading NPZ files
+- microsecond precision logging
+
+These additions have only been tested for TFRecord files, and PyTorch Dataloader with NPZ files as this is what our workload use. Because of this, other functionality is probably broken. More specifically, we're not trying to support the HDF5, CSV and NPZ readers, and we don't use Darshan or Tensorboard profiling.
+
+# Current Limitations
+
+There is an abtraction problem where DLIO considers PyTorch's data loader as a data type when it really is data type agnostic.
+Any of the existing data formats could be supported by extending the Dataset class and implementing the __len__ and __getitem__ methods. https://pytorch.org/docs/stable/data.html
+
+Since UNET3D reads in NPZ data, we currently only support this format for PyTorch. We will have to modify this for DLRM which reads in a custom binary format.
+
+Considering our objectives and the already significant changes made, I believe it might make more sense to change the abstraction to be workload based, instead of framework/datareader based.
```

Comment:
This makes sense. I like the abstraction of having a framework. My recommendation is to have the following abstraction:
- Framework (switches between tf and pytorch)
- DataReader (framework agnostic data loading using iterators and yields)
- FrameworkDataLoader (framework specific loading for DataReader)

What do you think?

Reply:
If I understand, we would have commands like `dlio --framework tf/pt --framework-dataloader y/n --format npz`. 

Depending on the `--framework-dataloader` flag, DLIO would use a TF Dataset, PT Dataloader, or a framework agnostic reader.

For file organization, we would keep the current `reader/` (minus `tf_reader.py` and `data_loader_reader.py`) and have a new folder like `framework_loader/` with a tf.data.Dataset-returning class and a torch.data.Dataloader-returning class, and ideally we would wrap the `reader/` classes inside.

I think this would be nice! However, I'm unsure of how easily we can wrap the `reader/` classes in the framework-specific data loaders. For example, the current `npz_reader` works great by itself, but to use it in a PyTorch Dataloader we would want to skip loading all the data in the `read()` method (to benefit from the on-demand loading aspect of data loaders), and thus couldn't simply wrap the `next()` method which yields the data assuming it has been read already. 

We could implement alternative paths etc, but the `npz_reader` class would then become complicated, when all that is needed to support npz in a PyTorch Dataloader is the 10-line NpzDataset we currently have in `data_loader_reader.py`.

Similarly, it might not be straight-forward to wrap `npz_reader` in a tf.data.Dataset.

So we may have to keep additional `pytorch_datasets/npz_dataset` and `tf_datasets/npz_dataset` classes, and do the same for each format we want to support.

Reply:
I was thinking in these lines.
- we can have a base reader which is actual calls to npz.
- If we want to change how its implemented on pytorch or tf, we can extend it and do the extensions appropriately. 

This way we have a cleaner abstraction. What do you think?